File: package/example/ios/Frame Processor Plugins/Example Plugin/ExampleFrameProcessorPlugin.m
//
//  ExampleFrameProcessorPlugin.m
//  VisionCameraExample
//
//  Created by Marc Rousavy on 01.05.21.
//

#if __has_include(<VisionCamera/FrameProcessorPlugin.h>)
#import <Foundation/Foundation.h>
#import <VisionCamera/FrameProcessorPlugin.h>
#import <VisionCamera/FrameProcessorPluginRegistry.h>
#import <VisionCamera/Frame.h>
#import <VisionCamera/SharedArray.h>

// Example for an Objective-C Frame Processor plugin

@interface ExampleFrameProcessorPlugin : FrameProcessorPlugin
@end

@implementation ExampleFrameProcessorPlugin {
  SharedArray* _sharedArray;
}

- (instancetype)initWithProxy:(VisionCameraProxyHolder*)proxy
                  withOptions:(NSDictionary* _Nullable)options {
  if (self = [super initWithProxy:proxy withOptions:options]) {
    _sharedArray = [[SharedArray alloc] initWithProxy:proxy
                                                 size:5];
    NSLog(@"ExampleFrameProcessorPlugin initialized with options: %@", options);
  }
  return self;
}

- (id)callback:(Frame *)frame withArguments:(NSDictionary *)arguments {
  CVPixelBufferRef imageBuffer = CMSampleBufferGetImageBuffer(frame.buffer);
  NSLog(@"ExamplePlugin: %zu x %zu Image. Logging %lu parameters:", CVPixelBufferGetWidth(imageBuffer), CVPixelBufferGetHeight(imageBuffer), (unsigned long)arguments.count);

  for (id param in arguments) {
    NSLog(@"ExamplePlugin:   -> %@ (%@)", param == nil ? @"(nil)" : [param description], NSStringFromClass([param classForCoder]));
  }

  uint8_t* data = _sharedArray.data;
  data[0] = (uint8_t)(random() * 100);

  return @{
    @"example_str": @"Test",
    @"example_bool": @(YES),
    @"example_double": @5.3,
    @"example_array": @[
      @"Hello",
      @(YES),
      @17.38
    ],
    @"example_array_buffer": _sharedArray
  };
}

VISION_EXPORT_FRAME_PROCESSOR(ExampleFrameProcessorPlugin, example_plugin)

@end
#endif

File: package/example/android/app/src/main/java/com/mrousavy/camera/example/ExampleFrameProcessorPlugin.java
package com.mrousavy.camera.example;

import android.media.Image;
import android.util.Log;

import com.mrousavy.camera.frameprocessor.Frame;
import com.mrousavy.camera.frameprocessor.FrameProcessorPlugin;
import com.mrousavy.camera.frameprocessor.SharedArray;
import com.mrousavy.camera.frameprocessor.VisionCameraProxy;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ExampleFrameProcessorPlugin extends FrameProcessorPlugin {
    SharedArray _sharedArray;
    private static final String TAG = "ExamplePlugin";

    @Override
    public Object callback(@NotNull Frame frame, @Nullable Map<String, Object> params) {
        if (params == null) return null;
        Image image = frame.getImage();

        Log.d(TAG, image.getWidth() + " x " + image.getHeight() + " Image with format #" + image.getFormat() + ". Logging " + params.size() + " parameters:");

        for (String key : params.keySet()) {
            Object value = params.get(key);
            Log.d(TAG, "  -> " + (value == null ? "(null)" : value + " (" + value.getClass().getName() + ")"));
        }

        Map<String, Object> map = new HashMap<>();
        map.put("example_str", "Test");
        map.put("example_bool", true);
        map.put("example_double", 5.3);

        List<Object> array = new ArrayList<>();
        array.add("Hello!");
        array.add(true);
        array.add(17.38);

        map.put("example_array", array);

        ByteBuffer byteBuffer = _sharedArray.getByteBuffer();
        byteBuffer.put(0, (byte)(Math.random() * 100));
        map.put("example_array_buffer", _sharedArray);

        return map;
    }

    ExampleFrameProcessorPlugin(VisionCameraProxy proxy, @Nullable Map<String, Object> options) {
        super();
        _sharedArray = new SharedArray(proxy, 5);
        Log.d(TAG, "Successfully allocated SharedArray! Size: " + _sharedArray.getSize());

        ByteBuffer buffer = ByteBuffer.allocateDirect(10);
        SharedArray testArray = new SharedArray(proxy, buffer);
        Log.d(TAG, "Successfully wrapped SharedArray in ByteBuffer! Size: " + testArray.getSize());

        Log.d(TAG, "ExampleFrameProcessorPlugin initialized with options: " + options);
    }
}
